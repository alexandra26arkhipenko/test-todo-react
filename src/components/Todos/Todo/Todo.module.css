.text{
    display:inline-block;
    margin-left: 15px;
    color:rgb(81 35 3);
}

.crossed{
    text-decoration: line-through;
    color:#d7cac1db
}

.delete {
    margin-left: auto;
    outline: none;
    color: #e8e8e8;
    display: none;
}

.delete:hover {
    color: rgb(114 12 12);
}

li:hover .delete {
    display: block;
}