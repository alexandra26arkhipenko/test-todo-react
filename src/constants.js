export const FOOTER_TEXT = 'Drag and drop to reorder list ';