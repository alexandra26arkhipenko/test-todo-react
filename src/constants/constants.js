export const FOOTER_TEXT = 'Drag and drop to reorder list ';

export const TODO_DELETE_BTN_TEXT = 'X';

export const TODO_INPUT_PLACEHOLDER = 'What are you going to do next?';

export const TODO_FOOTER_WELL_DONE = "Well done!";
export const TODO_FOOTER_ITEMS_LEFT = "items left";
export const TODO_FOOTER_FILTER_ALL = "All";
export const TODO_FILTER_ACTIVE_BTN = "Active";
export const TODO_FILTER_COMPLETED_BTN = "Completed";
