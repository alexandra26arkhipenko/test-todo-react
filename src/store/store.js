import { configureStore } from "@reduxjs/toolkit";
import { localStorageMiddleware } from "./middleware/localStorageMiddleware";

import todoReducer from "./todo";

const store = configureStore({
  reducer: todoReducer,
  middleware: [localStorageMiddleware],
});

export default store;
