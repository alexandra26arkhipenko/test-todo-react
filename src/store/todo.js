import { createSlice } from "@reduxjs/toolkit";
import { v4 as uuid } from "uuid";
import { getItemFromLocalStorage } from "../repository/localStorageRepository";
import { TODO_FOOTER_FILTER_ALL } from "../constants/constants";

const savedTodos = getItemFromLocalStorage("todos");
let todos = [];

if (savedTodos) {
  todos = JSON.parse(savedTodos);
}

const initialState = {
  todos: todos,
  filter: "",
};

const todoSlice = createSlice({
  name: "counter",
  initialState,
  reducers: {
    addTodo(state, action) {
      const newTodo = {
        id: uuid(),
        text: action.payload,
        isCompleted: false,
      };

      state.todos.unshift(newTodo);
    },
    compliteTodo(state, action) {
      const index = state.todos.findIndex((todo) => todo.id === action.payload);

      if (index !== -1) {
        state.todos[index].isCompleted = !state.todos[index].isCompleted;
      }
    },
    deleteTodo(state, action) {
      const index = state.todos.findIndex((todo) => todo.id === action.payload);

      if (index !== -1) {
        state.todos.splice(index, 1);
      }
    },
    deleteAllCompletedTodos(state) {
      state.todos = state.todos.filter((todo) => !todo.isCompleted);
    },
    filterTodos(state, action) {
      state.filter = action.payload;
    },
    reorderTodos(state, action) {
      let { dragIndex, dropIndex, filteredTodos } = action.payload;

      if (state.filterValue !== TODO_FOOTER_FILTER_ALL) {
        dragIndex = state.todos.findIndex(
          (property) => property.id === filteredTodos[dragIndex].id
        );
        dropIndex = state.todos.findIndex(
          (property) => property.id === filteredTodos[dropIndex].id
        );
      }

      const updatedTodos = [...state.todos];
      const [reorderedItem] = updatedTodos.splice(dragIndex, 1);
      updatedTodos.splice(dropIndex, 0, reorderedItem);
      state.todos = updatedTodos;
    },
  },
});

export const todoActions = todoSlice.actions;
export default todoSlice.reducer;
