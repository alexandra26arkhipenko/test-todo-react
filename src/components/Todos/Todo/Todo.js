import React from "react";
import PropTypes from "prop-types";
import { useDispatch } from "react-redux";
import { Draggable } from "react-beautiful-dnd";

import Button from "../../UI/Button/Button";
import Checkbox from "../../UI/Checkbox/Checkbox";
import { TODO_DELETE_BTN_TEXT } from "../../../constants/constants";
import { todoActions } from "../../../store/todo";

import classes from "./Todo.module.css";

const Todo = (props) => {
  const dispatch = useDispatch();

  const deleteHandler = () => {
    dispatch(todoActions.deleteTodo(props.id));
  };

  const checkboxHandler = () => {
    dispatch(todoActions.compliteTodo(props.id));
  };

  return (
    <Draggable key={props.id} draggableId={props.id} index={props.index}>
      {(provided) => (
        <li
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
        >
          <Checkbox checked={props.isCompleted} onChange={checkboxHandler} />
          <div
            className={`${classes.text} ${
              props.isCompleted && classes.crossed
            }`}
          >
            {props.text}
          </div>
          <Button
            type="button"
            className={classes.delete}
            onClick={deleteHandler}
          >
            {TODO_DELETE_BTN_TEXT}
          </Button>
        </li>
      )}
    </Draggable>
  );
};

Todo.propTypes = {
  id: PropTypes.string,
  index: PropTypes.number,
  isCompleted: PropTypes.bool,
  text: PropTypes.string,
};

export default Todo;
