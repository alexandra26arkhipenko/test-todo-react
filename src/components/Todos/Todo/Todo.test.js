import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Todo from "./Todo";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
import { todoActions } from "../../../store/todo";
import { TODO_DELETE_BTN_TEXT } from "../../../constants/constants";

const filteredTodos = [{ id: "1", text: "text", isCompleted: false }];
const mockStore = configureMockStore();
const store = mockStore({});

describe("Todo", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <DragDropContext>
          <Droppable droppableId="characters">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {filteredTodos.map((todo, index) => (
                  <Todo
                    key={todo.id}
                    id={todo.id}
                    text={todo.text}
                    index={index}
                    isCompleted={todo.isCompleted}
                  />
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should make todo complited when checkbox is clicked", () => {
    render(
        <Provider store={store}>
        <DragDropContext>
          <Droppable droppableId="characters">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {filteredTodos.map((todo, index) => (
                  <Todo
                    key={todo.id}
                    id={todo.id}
                    text={todo.text}
                    index={index}
                    isCompleted={todo.isCompleted}
                  />
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      </Provider>
    );    

    fireEvent.click(screen.getByRole('checkbox'));

    expect(store.getActions()).toContainEqual(todoActions.compliteTodo("1"));
  });

  it("should delete when delete button is clicked", () => {
    render(
        <Provider store={store}>
        <DragDropContext>
          <Droppable droppableId="characters">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {filteredTodos.map((todo, index) => (
                  <Todo
                    key={todo.id}
                    id={todo.id}
                    text={todo.text}
                    index={index}
                    isCompleted={todo.isCompleted}
                  />
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      </Provider>
    );    

    fireEvent.click(screen.getByText(TODO_DELETE_BTN_TEXT));

    expect(store.getActions()).toContainEqual(todoActions.deleteTodo("1"));
  });
});
