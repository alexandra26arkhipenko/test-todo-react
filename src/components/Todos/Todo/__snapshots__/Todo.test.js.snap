// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should render correctly 1`] = `
<div>
  <ul
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="characters"
  >
    <li
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="1"
      data-rbd-draggable-context-id="0"
      data-rbd-draggable-id="1"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <input
        class="round"
        type="checkbox"
      />
      <div
        class="text false"
      >
        text
      </div>
      <button
        class="button delete"
        type="button"
      >
        X
      </button>
    </li>
  </ul>
</div>
`;
